engine/rpcserver.go

```diff
@@ -1213,3 +1219,287 @@ func (s *RPCServer) GetAuditEvent(ctx context.Context, r *gctrpc.GetAuditEventRe
+	if !gctscript.GCTScriptConfig.Enabled {
+		return &gctrpc.GCTScriptGenericResponse{Status: gctscript.ErrScriptingDisabled.Error()}, nil
+	}
+
+	UUID, err := uuid.FromString(r.Script.UUID)
+	if err != nil {
+		return &gctrpc.GCTScriptGenericResponse{Status: MsgStatusError, Data: err.Error()}, nil
+	}
+
+	if v, f := gctscript.AllVMs[UUID]; f {
+		err = v.Shutdown()
+		status := " terminated"
+		if err != nil {
+			status = err.Error()
+		}
+		return &gctrpc.GCTScriptGenericResponse{Status: MsgStatusOK, Data: v.ID.String() + status}, nil
+	}
+	return &gctrpc.GCTScriptGenericResponse{Status: MsgStatusError, Data: "no running script found"}, nil
+}
+
+// GCTScriptUpload upload a new script to ScriptPath
+func (s *RPCServer) GCTScriptUpload(ctx context.Context, r *gctrpc.GCTScriptUploadRequest) (*gctrpc.GCTScriptGenericResponse, error) {
+	if !gctscript.GCTScriptConfig.Enabled {
+		return &gctrpc.GCTScriptGenericResponse{Status: gctscript.ErrScriptingDisabled.Error()}, nil
+	}
+
+	fPath := filepath.Join(gctscript.ScriptPath, r.ScriptName)
+	var fPathExits = fPath
+	if filepath.Ext(fPath) == ".zip" {
+		fPathExits = fPathExits[0 : len(fPathExits)-4]
+	}
+
+	if _, err := os.Stat(fPathExits); !os.IsNotExist(err) {
+		if !r.Overwrite {
+			return nil, fmt.Errorf("%s script found and overwrite set to false", r.ScriptName)
+		}
+		f := filepath.Join(gctscript.ScriptPath, "version_history")
+		err = os.MkdirAll(f, 0770)
+		if err != nil {
+			return nil, err
+		}
+		timeString := strconv.FormatInt(time.Now().UnixNano(), 10)
+		renamedFile := filepath.Join(f, timeString+"-"+filepath.Base(fPathExits))
+		err = os.Rename(fPathExits, renamedFile)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	newFile, err := os.Create(fPath)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = newFile.Write(r.Data)
+	if err != nil {
+		return nil, err
+	}
+	err = newFile.Close()
+	if err != nil {
+		log.Errorln(log.Global, "Failed to close file handle, archive removal may fail")
+	}
+
+	if r.Archived {
+		files, errExtract := extract.Unzip(fPath, filepath.Join(gctscript.ScriptPath, r.ScriptName[:len(r.ScriptName)-4]))
+		if errExtract != nil {
+			log.Errorf(log.Global, "Failed to extract zip file %v", errExtract)
+			return &gctrpc.GCTScriptGenericResponse{Status: MsgStatusError, Data: errExtract.Error()}, nil
+		}
+		for x := range files {
+			var failedFiles []string
+			err = gctscript.Validate(files[x])
+			if err != nil {
+				errRemove := os.Remove(fPath)
+				if errRemove != nil {
+					return nil, err
+				}
+				failedFiles = append(failedFiles, files[x])
+				return &gctrpc.GCTScriptGenericResponse{Status: ErrScriptFailedValidation, Data: strings.Join(failedFiles, ", ")}, nil
+			}
+		}
+		err = os.Remove(fPath)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		err = gctscript.Validate(fPath)
+		if err != nil {
+			errRemove := os.Remove(fPath)
+			if errRemove != nil {
+				return nil, err
```

Comment:
might want to return errRemove?

Reply:
looking over this again today i don't want to return here this early because it will not report why the file failed validation will rework the logic behind it